activejob/lib/active_job/execution.rb

```diff
@@ -38,7 +38,11 @@ def perform_now
       run_callbacks :perform do
         perform(*arguments)
       end
+
+      clear_lock
     rescue => exception
+      # ensure runs after rescue, which may re-enqueue the job, so we need to clear the lock first
+      clear_lock
```

Comment:
I think this open a race condition window. Between the lock is cleaned and the job is re-enqueued a new job can be enqueued getting the lock to itself and making our re-enqueue to fail. Should we care about this case? We solved this problem by checking if the job own the lock before the re-enqueue instead of cleaning the lock. We only clean the lock when the last retried is made.

Reply:
I'm not sure I follow the implementation you're proposing. Do you mean that in `enqueue` if the lock that has been acquired is for the current job you allow it through and you only clear the lock on successful completion, or when the job is exhausted (due to retries or errors)?

Reply:
The only problem that I see is the expectation that the job_id and retry counts are going to be different between the two jobs if the new job acquire the lock before the job that is retrying. What can cause the same job to be retried more than it should. I also not sure of the implication of this but to avoid that race condition we only release the lock after the last retry is done.

I'm fine with the current implementation but I think we should at least document this case.